File: core/src/main/kotlin/org/dreamexposure/discal/core/entities/Event.kt
package org.dreamexposure.discal.core.entities

import discord4j.common.util.Snowflake
import discord4j.core.`object`.entity.Guild
import kotlinx.serialization.encodeToString
import org.dreamexposure.discal.core.database.DatabaseManager
import org.dreamexposure.discal.core.entities.response.UpdateEventResponse
import org.dreamexposure.discal.core.entities.spec.update.UpdateEventSpec
import org.dreamexposure.discal.core.enums.announcement.AnnouncementType
import org.dreamexposure.discal.core.enums.event.EventColor
import org.dreamexposure.discal.core.`object`.announcement.Announcement
import org.dreamexposure.discal.core.`object`.event.EventData
import org.dreamexposure.discal.core.`object`.event.Recurrence
import org.dreamexposure.discal.core.`object`.event.RsvpData
import org.dreamexposure.discal.core.utils.GlobalVal.JSON_FORMAT
import org.json.JSONObject
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Duration
import java.time.Instant
import java.time.ZoneId
import java.time.temporal.ChronoUnit

interface Event {
    /**
     * The ID of the event.
     * In the format eXXXXXXXXXX if generated by DisCal, otherwise it was generated by the 3rd party calendar service
     */
    val eventId: String

    /**
     * The ID of the [Guild] this event belongs to.
     */
    val guildId: Snowflake
        get() = calendar.guildId

    /**
     * The [Calendar] this Event belongs to
     */
    val calendar: Calendar

    /**
     * The of the event saved to the DisCal database
     */
    val eventData: EventData

    /**
     * The name of the event, renamed from "summary" to make it more user-friendly and clear.
     */
    val name: String

    /**
     * A description of what the event is about.
     */
    val description: String

    /**
     * The location at which the event occurs, usually a map location.
     */
    val location: String

    /**
     * The link to view the event at
     */
    val link: String

    /**
     * The color of the event. Used for visually identifying it in Discord embeds.
     * If no event color is assigned, it returns [EventColor.NONE] which is DisCal blue.
     */
    val color: EventColor

    /**
     * The start of the event, as an [Instant] representing the time starting from January 1st 1970.
     */
    val start: Instant

    /**
     * The end of the event, as an [Instant] representing the time starting from January 1st 1970.
     */
    val end: Instant

    /**
     * Whether the event is a recurring event.
     */
    val recur: Boolean

    /**
     * The rules of the recurring event. Contains the RRule an human-readable information on how the event will recur
     */
    val recurrence: Recurrence

    /**
     * A link to the image, if none is present, returns empty
     */
    val image: String
        get() = eventData.imageLink

    /**
     * The timezone that the event takes place in. This is always the same as the [Calendar]'s timezone
     */
    val timezone: ZoneId
        get() = calendar.timezone

    //Reactive

    /**
     * Attempts to request the announcements linked to the event, such as a [SPECIFIC][AnnouncementType.SPECIFIC]
     * type announcement.
     * If an error occurs, it is emitted through the Flux.
     *
     * @return A [Flux] of all announcements that are linked to the event.
     */
    fun getLinkedAnnouncements(): Flux<Announcement> {
        return DatabaseManager.getAnnouncements(this.guildId)
            .flatMapMany { Flux.fromIterable(it) }
            .filter { ann ->
                when (ann.type) {
                    AnnouncementType.UNIVERSAL -> return@filter true
                    AnnouncementType.COLOR -> return@filter ann.eventColor == this.color
                    AnnouncementType.SPECIFIC -> return@filter ann.eventId == this.eventId
                    AnnouncementType.RECUR -> return@filter this.eventId.contains(ann.eventId)
                }
            }
    }

    /**
     * Attempts to request the [RsvpData] of the event.
     * If an error occurs, it is emitted through the Mono.
     *
     * @return A [Mono] containing the [RsvpData] of the event
     */
    fun getRsvp(): Mono<RsvpData> = DatabaseManager.getRsvpData(guildId, eventId)

    fun updateRsvp(rsvp: RsvpData) = DatabaseManager.updateRsvpData(rsvp)

    /**
     * Attempts to update the event and returns the result.
     * If an error occurs, it is emitted through the [Mono].
     *
     * @param spec The information to update the event with
     * @return A [Mono] that contains the [UpdateEventResponse] containing information on success and the changes.
     */
    fun update(spec: UpdateEventSpec): Mono<UpdateEventResponse>

    /**
     * Attempts to delete the event and returns the result.
     * If an error occurs, it is emitted through the [Mono].
     *
     * @return A [Mono] containing whether delete succeeded.
     */
    fun delete(): Mono<Boolean>

    fun isOngoing(): Boolean = start.isBefore(Instant.now()) && end.isAfter(Instant.now())

    fun isOver(): Boolean = end.isBefore(Instant.now())

    fun isStarted() = start.isBefore(Instant.now())

    /**
     * Whether the event is 24 hours long.
     *
     * @return Whether the event is 24 hours long
     */
    fun is24Hours() = Duration.between(start, end).toHours() == 24L

    /**
     * Whether the event lasts for a full calendar day (midnight to midnight) or longer.
     *
     * @return Whether the event is all day
     */
    fun isAllDay(): Boolean {
        val start = this.start.atZone(timezone)

        return start.hour == 0 && is24Hours()
    }

    /**
     * Whether the event spans across multiple calendar days (ex Monday 8pm to Tuesday 3am)
     *
     * @return Whether the event is multi-day
     */
    fun isMultiDay(): Boolean {
        if (isAllDay()) return false // All day events should not count as multi-day events

        val start = this.start.atZone(timezone).truncatedTo(ChronoUnit.DAYS)
        val end = this.end.atZone(timezone).truncatedTo(ChronoUnit.DAYS)

        return when {
            start.year != end.year -> true
            start.month != end.month -> true
            start.dayOfYear != end.dayOfYear -> true
            else -> false
        }
    }

    //Json bullshit
    fun toJson(): JSONObject {
        return JSONObject()
            .put("guild_id", guildId)
            .put("calendar", calendar.toJson())
            .put("event_id", eventId)
            .put("epoch_start", start.toEpochMilli())
            .put("epoch_end", end.toEpochMilli())
            .put("name", name)
            .put("description", description)
            .put("location", location)
            .put("is_parent", !eventId.contains("_"))
            .put("color", color.name)
            .put("recur", recur)
            .put("recurrence", JSONObject(JSON_FORMAT.encodeToString(recurrence)))
            .put("rrule", recurrence.toRRule())
            .put("image", eventData.imageLink)
    }
}

File: web/src/main/javascript/objects/event/Event.ts
import {EventColor, eventColorRGB} from "@/enums/EventColor";
import {Recurrence} from "@/objects/event/Recurrence";

export class Event {
    private _eventId: string = "";
    private _epochStart: number = 0;
    private _epochEnd: number = 0;

    private _summary: string = "";
    private _description: string = "";
    private _location: string = "";

    private _isParent: boolean = false;
    private _color: EventColor = EventColor.NONE;

    private _recur: boolean = false;
    private _recurrence: Recurrence = new Recurrence();
    private _rrule: String = "";

    private _image: string = "";

    constructor() {
    }

    //Setter/getter pairs
    get eventId() {
        return this._eventId;
    }

    set eventId(id) {
        this._eventId = id;
    }

    get epochStart() {
        return this._epochStart;
    }

    set epochStart(epoch) {
        this._epochStart = epoch;
    }

    get epochEnd() {
        return this._epochEnd;
    }

    set epochEnd(epoch) {
        this._epochEnd = epoch;
    }

    get summary() {
        return this._summary;
    }

    set summary(summary) {
        this._summary = summary;
    }

    get description() {
        return this._description;
    }

    set description(desc) {
        this._description = desc;
    }

    get location() {
        return this._location;
    }

    set location(loc) {
        this._location = loc;
    }

    get isParent() {
        return this._isParent;
    }

    set isParent(isPar) {
        this._isParent = isPar;
    }

    get color() {
        return this._color;
    }

    set color(col) {
        this._color = col;
    }

    get doesRecur() {
        return this._recur;
    }

    set doesRecur(does) {
        this._recur = does;
    }

    get recurrence() {
        return this._recurrence;
    }

    set recurrence(rec) {
        this._recurrence = rec;
    }

    get rrule() {
        return this._rrule;
    }

    set rrule(rr) {
        this._rrule = rr;
    }

    get image() {
        return this._image;
    }

    set image(img) {
        this._image = img;
    }


    //Json conversion
    toFullCalEvent(editable: boolean) {
        let event: any = {
            id: this.eventId,
            groupId: this.eventId.split("_")[0],
            title: this.summary,
            description: this.description,
            location: this.location,
            image: this.image,
            eventColor: this.color,

            start: this.epochStart,
            end: this.epochEnd,

            overlap: true,
            editable: editable,

            rawEvent: this,
        };

        if (this.color != EventColor.NONE) {
            event.borderColor = eventColorRGB(this.color);
        }
        if (this.doesRecur) {
            event.rrule = this.rrule;
        }

        return event;
    }


    toJson() {
        let json: any = {
            "event_id": this.eventId,
            "epoch_start": this.epochStart,
            "epoch_end": this.epochEnd,
            "color": EventColor[this.color],
            "recur": this.doesRecur
        };

        if (this.summary.length > 0) {
            json.name = this.summary;
        }
        if (this.description.length) {
            json.description = this.description;
        }
        if (this.location.length > 0) {
            json.location = this.location;
        }
        if (this.doesRecur) {
            json.recurrence = this.recurrence.toJson();
            json.rrule = this.rrule
        }
        if (this.image.length > 0) {
            json.image = this.image;
        }

        return json;
    }

    fromJson(json: any) {
        this.eventId = json.event_id;
        this.epochStart = json.epoch_start;
        this.epochEnd = json.epoch_end;

        if (json.hasOwnProperty("name")) {
            this.summary = json.name;
        }
        if (json.hasOwnProperty("description")) {
            this.description = json.description;
        }
        if (json.hasOwnProperty("location")) {
            this.location = json.location;
        }

        this.isParent = json.is_parent;
        this.color = (<any>EventColor)[json.color];

        this.doesRecur = json.recur;
        if (this.doesRecur) {
            this.recurrence = new Recurrence().fromJson(json.recurrence);
            this.rrule = json.rrule;
        }

        if (json.hasOwnProperty("image")) {
            this.image = json.image;
        }

        return this;
    }
}
